internal/app/server/impl.go

```diff
@@ -0,0 +1,361 @@
+)
+
+// ConformanceRequest is a general interface for all conformance requests (UnaryRequest, ServerStreamRequest, etc.)
+type ConformanceRequest interface {
+	GetResponseHeaders() []*v1alpha1.Header
+	GetResponseTrailers() []*v1alpha1.Header
+}
+
+// Stream represents a stream with the ability to set headers and trailers and send a message T
+type Stream[T any] interface {
+	ResponseHeader() http.Header
+	ResponseTrailer() http.Header
+	Send(*T) error
+}
+
+type conformanceServer struct{}
+
+func (s *conformanceServer) Unary(
+	ctx context.Context,
+	req *connect.Request[v1alpha1.UnaryRequest],
+) (*connect.Response[v1alpha1.UnaryResponse], error) {
+	msgAsAny, err := asAny(req.Msg)
+	if err != nil {
+		return nil, err
+	}
+	payload, err := parseUnaryResponseDefinition(req.Msg.ResponseDefinition, req.Header(), []*anypb.Any{msgAsAny})
+	if err != nil {
+		return nil, err
+	}
+
+	resp := connect.NewResponse(&v1alpha1.UnaryResponse{
+		Payload: payload,
+	})
+
+	if req.Msg.ResponseDefinition != nil {
+		setResponseHeadersAndTrailers(req.Msg.ResponseDefinition, resp)
+	}
+
+	return resp, nil
+}
+
+func (s *conformanceServer) ClientStream(
+	ctx context.Context,
+	stream *connect.ClientStream[v1alpha1.ClientStreamRequest],
+) (*connect.Response[v1alpha1.ClientStreamResponse], error) {
+	var responseDefinition *v1alpha1.UnaryResponseDefinition
+	firstRecv := true
+	var reqs []*anypb.Any
+	for stream.Receive() {
+		if err := ctx.Err(); err != nil {
+			return nil, err
+		}
+		msg := stream.Msg()
+		// If this is the first message received on the stream, save off the response definition we need to send
+		if firstRecv {
+			responseDefinition = msg.ResponseDefinition
+			firstRecv = false
+		}
+		// Record all the requests received
+		msgAsAny, err := asAny(msg)
+		if err != nil {
+			return nil, err
+		}
+		reqs = append(reqs, msgAsAny)
+	}
+	if err := stream.Err(); err != nil {
+		return nil, err
+	}
+
+	payload, err := parseUnaryResponseDefinition(responseDefinition, stream.RequestHeader(), reqs)
+
+	resp := connect.NewResponse(&v1alpha1.ClientStreamResponse{
+		Payload: payload,
+	})
+
+	if responseDefinition != nil {
+		setResponseHeadersAndTrailers(responseDefinition, resp)
+	}
+
+	return resp, err
+}
+
+func (s *conformanceServer) ServerStream(
+	ctx context.Context,
+	req *connect.Request[v1alpha1.ServerStreamRequest],
+	stream *connect.ServerStream[v1alpha1.ServerStreamResponse],
+) error {
+	responseDefinition := req.Msg.ResponseDefinition
+	if responseDefinition != nil {
+		// Set all requested response headers on the response
+		for _, header := range responseDefinition.ResponseHeaders {
+			for _, val := range header.Value {
+				stream.ResponseHeader().Add(header.Name, val)
+			}
+		}
+		// Set all requested response trailers on the response
+		for _, trailer := range responseDefinition.ResponseTrailers {
+			for _, val := range trailer.Value {
+				stream.ResponseTrailer().Add(trailer.Name, val)
+			}
+		}
```

Comment:
I don't like that this block of code is repeated a few times. There's just a bit of difficulty abstracting it with the APIs for `stream.ResponseTrailer()` and `resp.Trailer()`. We could definitely do it, but I'm not sure if it just makes it overly confusing vs. simply repeating the logic.